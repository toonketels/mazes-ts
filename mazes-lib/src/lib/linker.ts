/**
 * Link two cells together
 */
import * as _ from 'lodash';
import {Cell} from "./grid";

export function link(a: Cell, b: Cell) {

    // @TODO: maybe a currentCell is defined by its position in the grid
    //        so it has some kind of identity
    //        maybe it should know if its being linked to itself
    if (_.isEqual(a, b)) throw new Error("Cannot link a currentCell to itself")

    if (!a.links.includes(b)) a.links = [...a.links, b]
    if (!b.links.includes(a)) b.links = [...b.links, a]
}

export function unlink(a: Cell, b: Cell) {
    if (a.links.includes(b)) a.links = _.without(a.links, b)
    if (b.links.includes(a)) b.links = _.without(b.links, a)
}