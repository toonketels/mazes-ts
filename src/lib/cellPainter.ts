import {Cell} from "./grid";

export interface CellPaintOptions {
    omitNorth?: boolean;
    omitEast?: boolean;
    omitSouth?: boolean;
    omitWest?: boolean;
}

export interface CellPainter {
    (cell: Cell, options?: CellPaintOptions): string[];
}

export type Direction = "north" | "east" | "south" | "west";

export const coordinatePainer: CellPainter = cell => {
    return [` ${cell.x}${cell.y} `]
}
export const wallPainter: CellPainter = (cell, {
    omitNorth,
    omitEast,
    omitSouth,
    omitWest
} = {}) => {

    const north = isLinked(cell, "north") ? `+   +` : `+---+`;
    const east = isLinked(cell, "east") ? ` ` : `|`;
    const south = isLinked(cell, "south") ? `+   +` : `+---+`;
    const west = isLinked(cell, "west") ? ` ` : `|`;

    const start = omitWest ? 1 : 0
    const end = omitEast ? 4 : 5

    return [
        ...(omitNorth ? [] : [north.substring(start, end)]),
        `${west}   ${east}`.substring(start, end),
        ...(omitSouth ? [] : [south.substring(start, end)])
    ]
}


export function isLinked(cell: Cell, direction: Direction): boolean {
    switch(direction) {
        case "north":
            return !!cell.links.find(({y}) => y === cell.y - 1 )
        case "east":
            return !!cell.links.find(({x}) => x === cell.x + 1 )
        case "south":
            return !!cell.links.find(({y}) => y === cell.y + 1 )
        case "west":
            return !!cell.links.find(({x}) => x === cell.x - 1 )
        default:
            const checkExhaustive: never = direction
            return checkExhaustive
    }
}