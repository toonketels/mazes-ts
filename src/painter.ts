import {Cell} from "./grid";

export interface Painter {
    (cell: Cell): string[];
}

export function coordinatePainer(cell: Cell): string[] {
    return [` ${cell.x}${cell.y} `]
}

export function wallPainter(cell: Cell): string[] {
    const north = isLinked(cell, "north") ? `+   +` : `+---+`;
    const east = isLinked(cell, "east") ? ` ` : `|`;
    const south = isLinked(cell, "south") ? `+   +` : `+---+`;
    const west = isLinked(cell, "west") ? ` ` : `|`;

    return [
        north,
        `${west}   ${east}`,
        south
    ]

    function isLinked(cell: Cell, direction: "north"| "east" | "south" | "west"): boolean {
        switch(direction) {
            case "north":
                return !!cell.links.find(({y}) => y === cell.y - 1 )
            case "east":
                return !!cell.links.find(({x}) => x === cell.x + 1 )
            case "south":
                return !!cell.links.find(({y}) => y === cell.y + 1 )
            case "west":
                return !!cell.links.find(({x}) => x === cell.x - 1 )
            default:
                const checkExhaustive: never = direction
                return checkExhaustive
        }
    }
}
