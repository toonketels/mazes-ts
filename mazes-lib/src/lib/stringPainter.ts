import {Cell, Grid} from "./grid";
import {coordinatePainer, CellPainter} from "./cellPainter";
import {Painter} from "../index";

export type StringPainter = Painter<string>

export interface StringPainterFactoryOptions {
    cellPainter?: CellPainter
}

export function createStringPainter({cellPainter = coordinatePainer}: StringPainterFactoryOptions = {}): StringPainter {
    return (grid) => stringPainter(grid, cellPainter)
}

function stringPainter(grid: Grid, painter: CellPainter): string {
    let result = ""

    for (let [y, row] of grid.rowsIndexed()) {
        let jobs = []
        for (let [x, cell] of row) {

            const isLastRow = y === grid.dimensions.height - 1
            const isLastColumn = x === grid.dimensions.width - 1

            const ops = {
                omitSouth: !isLastRow,
                omitEast: !isLastColumn
            }

            jobs.push(painter(cell, ops))
        }

        if (jobs[0] === undefined) return "\n"
        const linesPerJob = jobs[0].length


        for (let line = 0; line < linesPerJob; line++) {
            for (let job = 0; job < jobs.length; job++) {
                result += jobs[job][line]
            }
            result += "\n"
        }
    }

    return result
}