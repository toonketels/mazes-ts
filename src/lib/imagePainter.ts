import { PNG } from 'pngjs'
import * as fs from 'fs'
import {Cell, Grid} from "./grid";
import {isLinked} from "./cellPainter";
import {Painter} from "./lib";

export interface ImagePaintResult {
    filePath: string
}

export type ImagePainter = Painter<ImagePaintResult>

export interface ImagePainterFactoryOptions {
    directory: string,
    fileName: string
}

export function createImagePainter({directory, fileName}: ImagePainterFactoryOptions): ImagePainter {

    // maybe a bit strange to make the filename part of the factory
    let counter = 0;

    return function paintGridToImage(grid: Grid): ImagePaintResult {


        const cellDimension = 20
        const wallWidth = 4
        const alpha = 255;
        const numberOfChannels = 4;

        const newfile = new PNG({
            height: getTotalImageDimension(grid.dimensions.height),
            width: getTotalImageDimension(grid.dimensions.width),
            fill: true,
            bgColor: { red: 255, green: 255, blue: 255}
        })

        function getTotalImageDimension(cellCount: number) {
            return (cellCount * (cellDimension + wallWidth)) + wallWidth;

        }
        function pixelInScanline(coordinate: ImageCoordinate): number {
            const scanline = coordinate.y
            const pixel = coordinate.x
            return (newfile.width * scanline + pixel) * numberOfChannels;
        }

        function scaleToImg(cellCoordinate: number) {
            const imageCoordinate = cellCoordinate * (wallWidth + cellDimension)
            return imageCoordinate
        }

        function scaleToImageCoordinate(x: number, y: number, fn: (c: ImageCoordinate) => ImageCoordinate = (i) => i): ImageCoordinate {
            const coordinate = new ImageCoordinate(scaleToImg(x), scaleToImg(y))
            return fn(coordinate)
        }

        function paint(coordinate: ImageCoordinate, r: number, g: number, b: number, a: number) {
            const refImageIndex = pixelInScanline(coordinate)
            newfile.data[refImageIndex] = r
            newfile.data[refImageIndex + 1] = g
            newfile.data[refImageIndex + 2] = b

            newfile.data[refImageIndex + 3] = a
        }
        function paintWithColor(coordinate: ImageCoordinate, color: "black" | "white") {
            switch (color) {
                case "black": paint(coordinate, 0, 0, 0, alpha)
                    break;
                case "white": paint(coordinate, 255, 255, 255, alpha)
                    break;
                default:
                    const checkExhaustive: never = color
                    break;
            }
        }

        function paintRectangle({x: x1, y: y1}: ImageCoordinate, {x: x2, y: y2}: ImageCoordinate, color: "black" | "white" = "black") {

            if (x1 > x2) throw Error("first coordinate x cannot be bigger")
            if (y1 > y2) throw Error("first coordinate y cannot be bigger")

            for (let y = y1; y <= y2; y++) {
                for (let x = x1; x <= x2; x++) {
                    paintWithColor(new ImageCoordinate(x, y), color)
                }
            }



        }

        function paintCell(cell: Cell) {
            const { x, y } = cell

            paintRectangle(scaleToImageCoordinate(x, y),
                scaleToImageCoordinate(x, y, c => {
                    c.x = c.x + wallWidth + cellDimension - 1
                    c.y = c.y + wallWidth + cellDimension - 1
                    return c
                }), "white")

            if (!isLinked(cell, "north")) {
                paintRectangle(scaleToImageCoordinate(x, y), scaleToImageCoordinate(x, y, c => {
                    c.x = c.x + wallWidth + wallWidth + cellDimension - 1
                    c.y = c.y + wallWidth - 1
                    return c
                }))

            }
            if (!isLinked(cell, "west")) {
                paintRectangle(
                    scaleToImageCoordinate(x, y),
                    scaleToImageCoordinate(x, y, c => {
                        c.x = c.x + wallWidth - 1
                        c.y = c.y + wallWidth + wallWidth + cellDimension - 1
                        return c
                    }))

            }
        }

        for (let row of grid.rows()) {
            for (let cell of row) {
                paintCell(cell)
            }
        }

        // Paint east wall
        paintRectangle(scaleToImageCoordinate(grid.dimensions.width, 0), scaleToImageCoordinate(grid.dimensions.width, grid.dimensions.height, c => {
            c.x = c.x + wallWidth - 1
            c.y = c.y + wallWidth - 1
            return c
        }))

        // Paing south wall
        paintRectangle(scaleToImageCoordinate(0, grid.dimensions.height), scaleToImageCoordinate(grid.dimensions.width, grid.dimensions.height, c => {
            c.x = c.x + wallWidth - 1
            c.y = c.y + wallWidth - 1
            return c
        }))

        const name = counter === 0 ? fileName : `${fileName}-${counter}`
        counter++
        let filePath = directory + `/${name}.png`;
        //
        // newfile
        //     .pack()
        //     .pipe(fs.createWriteStream(filePath))
        //     .on("finish", (...args) => { console.log("done", ...args) })

        var buffer = PNG.sync.write(newfile);
        fs.writeFileSync(filePath, buffer);

        return { filePath }

    }
}



class ImageCoordinate { constructor (public x: number, public y: number) {} }