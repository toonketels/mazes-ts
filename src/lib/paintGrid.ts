import {Grid} from "./grid";
import {coordinatePainer, Painter} from "./painter";

export function paintGrid(grid: Grid, painter: Painter): string {
    let result = ""

    for (let [y, row] of grid.rowsIndexed()) {
        let jobs = []
        for (let [x, cell] of row) {

            const isLastRow = y === grid.dimensions.height - 1
            const isLastColumn = x === grid.dimensions.width - 1

            const ops = {
                omitSouth: !isLastRow,
                omitEast: !isLastColumn
            }

            jobs.push(painter(cell, ops))
        }

        if (jobs[0] === undefined) return "\n"
        const linesPerJob = jobs[0].length


        for (let line = 0; line < linesPerJob; line++) {
            for (let job = 0; job < jobs.length; job++) {
                result += jobs[job][line]
            }
            result += "\n"
        }
    }

    return result
}