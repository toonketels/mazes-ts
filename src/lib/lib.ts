import {CreateGrid, createGrid as createAGrid, Dimensions} from "./grid";
import {coordinatePainer, Painter} from "./painter";
import {createBinaryTreeMazeBuilder, CreateMazeBuilder} from "./mazeBuilder";
import {paintGrid} from "./paintGrid";

interface PaintMazeOptions {
    dimensions: Dimensions,
    seed?: string,
    createGrid?: CreateGrid,
    createMazeBuilder?: CreateMazeBuilder,
    painter?: Painter
}

type CreateMazeOptions = Pick<PaintMazeOptions, "dimensions" | "seed" | "createGrid" | "createMazeBuilder">


export function paintMaze({
                               dimensions,
                               seed,
                               createGrid,
                               createMazeBuilder,
                               painter = coordinatePainer
}: PaintMazeOptions): string {

    let {grid, mazeBuilder} = initMaze({createGrid, dimensions, createMazeBuilder, seed});

    mazeBuilder.build()

    return paintGrid(grid, painter)
}

export function* paintMazeCreation({
                                      dimensions,
                                      seed,
                                      createGrid = createAGrid,
                                      createMazeBuilder = createBinaryTreeMazeBuilder,
                                      painter = coordinatePainer
                                  }: PaintMazeOptions): Generator<string> {

    let grid = createGrid({dimensions});
    let mazeBuilder = createMazeBuilder({grid, seed});


    for (let grid of mazeBuilder) {
        yield paintGrid(grid, painter)
    }
}

function initMaze({
                      dimensions,
                      seed,
                      createGrid = createAGrid,
                      createMazeBuilder = createBinaryTreeMazeBuilder
                  }: CreateMazeOptions) {
    let grid = createGrid({dimensions});
    let mazeBuilder = createMazeBuilder({grid, seed});
    return {grid, mazeBuilder};
}
