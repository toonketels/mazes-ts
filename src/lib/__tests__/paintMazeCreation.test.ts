import {wallPainter} from "../painter";
import {paintMazeCreation} from "../lib";

describe("paintMazeCreation()", () => {

    test("paints every step of maze creation", () => {

        const frames = paintMazeCreation({
            dimensions: {width: 3, height: 3},
            seed: "a seed for predictable randomness",
            painter: wallPainter});

        for(let frame of frames) {
            expect(frame).toMatchSnapshot()
        }
    })
})