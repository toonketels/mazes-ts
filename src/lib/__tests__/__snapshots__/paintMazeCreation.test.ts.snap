// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paintMazeCreation() paints every step of maze creation 1`] = `
"+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 2`] = `
"+---+---+---+
|   |   |   |
+   +---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 3`] = `
"+---+---+---+
|   |       |
+   +---+---+
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 4`] = `
"+---+---+---+
|   |       |
+   +---+   +
|   |   |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 5`] = `
"+---+---+---+
|   |       |
+   +---+   +
|       |   |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 6`] = `
"+---+---+---+
|   |       |
+   +---+   +
|           |
+---+---+---+
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 7`] = `
"+---+---+---+
|   |       |
+   +---+   +
|           |
+---+---+   +
|   |   |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 8`] = `
"+---+---+---+
|   |       |
+   +---+   +
|           |
+---+---+   +
|       |   |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 9`] = `
"+---+---+---+
|   |       |
+   +---+   +
|           |
+---+---+   +
|           |
+---+---+---+
"
`;

exports[`paintMazeCreation() paints every step of maze creation 10`] = `
"+---+---+---+
|   |       |
+   +---+   +
|           |
+---+---+   +
|           |
+---+---+---+
"
`;
